File: skincare-routines/src/chat/local/tools/inciMatch.ts
/** Ingredient matching for the product tools — twin of scripts/lib/inci-aliases.cjs (build) and
 * chat_api/tools/inci_match.py. Query terms go through the same normalisation as the indexed INCI columns, then
 * widen through the alias table ("iron oxide" → CI 77491/77492/77499) and are matched as whole names. */
import type { IngredientAlias } from '../../../lib/types';

export function normalizeInci(text: string): string {
  return text
    .normalize('NFKD')
    .replace(/[\u0300-\u036f]/g, '')
    .toLowerCase()
    .replace(/\bc\.?\s?i\.?\s*(?:no\.?)?\s*[-:.]?\s*(\d{5})\b/g, 'ci $1')
    .replace(/\(\s*and\s*\)|[,;/|\u2022\u00b7]|\s\band\b\s/g, ',')
    .replace(/[^a-z0-9,]+/g, ' ')
    .replace(/\s*,\s*/g, ',')
    .replace(/,+/g, ',')
    .replace(/^,|,$/g, '')
    .trim();
}

export interface IngredientQuery {
  /** The user's term as typed. */
  term: string;
  /** Alias-table label when the term was recognised (e.g. "Iron oxides (tint pigments)"), else null. */
  label: string | null;
  /** Names matched anywhere inside an ingredient. */
  inci: string[];
  /** Names that must be the entire ingredient ("alcohol", not "cetearyl alcohol"). */
  whole: string[];
  patterns: RegExp[];
  wholePatterns: RegExp[];
}

const escapeRe = (s: string) => s.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');

/** Entries with this many INCI names or fewer are one substance under a few spellings; larger ones are groups
 * (mineral filters, retinoids) that a single INCI name must never widen into. Same constant as inci-aliases.cjs. */
const SYNONYM_ENTRY_MAX = 3;

function resolveAliasEntries(t: string, aliases: IngredientAlias[]): IngredientAlias[] {
  const own = aliases.filter((e) => e.aliases.includes(t) || e.whole.includes(t) || e.id.toLowerCase() === t);
  if (own.length) return own;
  return aliases.filter((e) => e.inci.includes(t) && e.inci.length <= SYNONYM_ENTRY_MAX);
}

export function expandIngredient(term: string, aliases: IngredientAlias[]): IngredientQuery {
  const t = normalizeInci(term);
  const inci = new Set<string>();
  const whole = new Set<string>();
  const labels: string[] = [];
  for (const entry of resolveAliasEntries(t, aliases)) {
    entry.inci.forEach((n) => inci.add(n));
    entry.whole.forEach((n) => whole.add(n));
    labels.push(entry.label);
  }
  if (t && !whole.has(t) && !labels.length) inci.add(t);
  const names = [...inci];
  const wholeNames = [...whole];
  return {
    term, label: labels[0] ?? null, inci: names, whole: wholeNames,
    patterns: names.map((n) => new RegExp(`(?:^|[ ,])${escapeRe(n)}(?:s|es)?(?:[ ,]|$)`)),
    wholePatterns: wholeNames.map((n) => new RegExp(`(?:^|,)${escapeRe(n)}(?:,|$)`)),
  };
}

/** The first name of `q` found in normalised INCI text, or null. */
export function findIngredient(text: string, q: IngredientQuery): string | null {
  for (let i = 0; i < q.patterns.length; i++) if (q.patterns[i].test(text)) return q.inci[i];
  for (let i = 0; i < q.wholePatterns.length; i++) if (q.wholePatterns[i].test(text)) return q.whole[i];
  return null;
}

/** Brand + title words, folded the same way for `title_words` constraints (prefix match per word). */
export const titleTokens = (text: string) => normalizeInci(text).replace(/,/g, ' ').split(' ').filter(Boolean);

export function titleHasWords(tokens: string[], words: string[]): boolean {
  return words.every((w) => {
    const parts = titleTokens(w);
    return parts.every((p) => tokens.some((tok) => tok === p || tok.startsWith(p)));
  });
}

File: skincare-routines/chat-api/chat_api/tools/inci_match.py
"""Ingredient matching for the product tools — twin of scripts/lib/inci-aliases.cjs and src/chat/local/tools/inciMatch.ts.

Query terms get the same normalisation as the indexed INCI columns, widen through the alias table
("iron oxide" → CI 77491/77492/77499) and are matched as whole ingredient names."""
from __future__ import annotations

import re
import unicodedata
from dataclasses import dataclass, field

_CI = re.compile(r"\bc\.?\s?i\.?\s*(?:no\.?)?\s*[-:.]?\s*(\d{5})\b")
_SEP = re.compile(r"\(\s*and\s*\)|[,;/|\u2022\u00b7]|\s\band\b\s")
_JUNK = re.compile(r"[^a-z0-9,]+")


def normalize_inci(text: str) -> str:
    t = unicodedata.normalize("NFKD", text or "")
    t = "".join(ch for ch in t if not unicodedata.combining(ch)).lower()
    t = _CI.sub(r"ci \1", t)
    t = _SEP.sub(",", t)
    t = _JUNK.sub(" ", t)
    t = re.sub(r"\s*,\s*", ",", t)
    t = re.sub(r",+", ",", t)
    return t.strip(",").strip()


@dataclass
class IngredientQuery:
    term: str
    label: str | None
    inci: list[str]
    whole: list[str]
    patterns: list[re.Pattern[str]] = field(default_factory=list)
    whole_patterns: list[re.Pattern[str]] = field(default_factory=list)

    def looking_for(self) -> list[str]:
        return [*self.inci, *self.whole]


# Entries with this many INCI names or fewer are one substance under a few spellings; larger ones are groups
# (mineral filters, retinoids) that a single INCI name must never widen into. Same constant as inci-aliases.cjs.
SYNONYM_ENTRY_MAX = 3


def resolve_alias_entries(t: str, aliases: list[dict]) -> list[dict]:
    own = [e for e in aliases if t in e.get("aliases", []) or t in e.get("whole", []) or str(e.get("id", "")).lower() == t]
    if own:
        return own
    return [e for e in aliases if t in e.get("inci", []) and len(e.get("inci", [])) <= SYNONYM_ENTRY_MAX]


def expand_ingredient(term: str, aliases: list[dict]) -> IngredientQuery:
    t = normalize_inci(term)
    inci: dict[str, None] = {}
    whole: dict[str, None] = {}
    labels: list[str] = []
    for entry in resolve_alias_entries(t, aliases):
        for n in entry.get("inci", []):
            inci[n] = None
        for n in entry.get("whole", []):
            whole[n] = None
        labels.append(str(entry.get("label", "")))
    if t and t not in whole and not labels:
        inci[t] = None
    names = list(inci)
    whole_names = list(whole)
    return IngredientQuery(
        term=term,
        label=labels[0] if labels else None,
        inci=names,
        whole=whole_names,
        patterns=[re.compile(rf"(?:^|[ ,]){re.escape(n)}(?:s|es)?(?:[ ,]|$)") for n in names],
        whole_patterns=[re.compile(rf"(?:^|,){re.escape(n)}(?:,|$)") for n in whole_names],
    )


def find_ingredient(text: str, q: IngredientQuery) -> str | None:
    """The first name of `q` found in normalised INCI text, or None."""
    for name, pat in zip(q.inci, q.patterns, strict=True):
        if pat.search(text):
            return name
    for name, pat in zip(q.whole, q.whole_patterns, strict=True):
        if pat.search(text):
            return name
    return None


def title_tokens(text: str) -> list[str]:
    return [tok for tok in normalize_inci(text).replace(",", " ").split(" ") if tok]


def title_has_words(tokens: list[str], words: list[str]) -> bool:
    """Every word (prefix match per token) appears in the folded brand + title."""
    return all(all(any(tok == p or tok.startswith(p) for tok in tokens) for p in title_tokens(w)) for w in words)
